### PR TITLE
[SPARK-43229][ML][PYTHON][CONNECT] Introduce Barrier Python UDF

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

This PR reverts changes from [SPARK-42896](https://issues.apache.org/jira/browse/SPARK-42896) and [SPARK-42929](https://issues.apache.org/jira/browse/SPARK-42929) which made `mapInPandas` and `mapInArrow` support barrier mode execution via adding a `is_barrier` parameter.

Instead, this PR add `is_barrier` as a optional function attribute, so that the barrier python UDF can be supported in physical operators other than `MapInBatchExec`.

Right now I want to narrow its usage scope because it is only used in ML:

- only supported in Pandas UDF;
- only supported in `MapInPandas` and `MapInArrow`;
- can not be registered;

This PR will not add a user-facing API or Parameter or Annotation, instead only a _private_ function attribute will be added, that is because:

- Right now it is only needed to integrate external ML training frameworks like PyTorch and XGBoost, which just run `collect` after the UDF execution to obtain the model coefficients;
- The limitation of existing barrier mode: many RDD operations are not supported, see [1](https://github.com/apache/spark/blob/1a6b1770c85f37982b15d261abf9cc6e4be740f4/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L474-L488) and [2](https://github.com/apache/spark/blob/485145326a9c97ede260b0e267ee116f182cfd56/core/src/main/scala/org/apache/spark/scheduler/BarrierJobAllocationFailed.scala#L42-L65) . And it is non-trivial to make `RDDBarrier` fully compatible with `RDD`.
 
A simple example to illustrate this problem:
```
In [1]: df = spark.createDataFrame([(1, 21), (2, 30)], ("id", "age"))

In [2]: def filter_func(iterator):
   ...:     for pdf in iterator:
   ...:         yield pdf[pdf.id == 1]
   ...: 

In [3]: df.mapInPandas(filter_func, df.schema).explain()
== Physical Plan ==
MapInPandas filter_func(id#0L, age#1L)#4, [id#5L, age#6L], false
+- *(1) Scan ExistingRDD[id#0L,age#1L]



In [4]: df.mapInPandas(filter_func, df.schema).show()
+---+---+                                                                       
| id|age|
+---+---+
|  1| 21|
+---+---+


In [5]: df.mapInPandas(filter_func, df.schema, barrier=True).explain()
== Physical Plan ==
MapInPandas filter_func(id#0L, age#1L)#23, [id#24L, age#25L], true
+- *(1) Scan ExistingRDD[id#0L,age#1L]



In [6]: df.mapInPandas(filter_func, df.schema, barrier=True).show()
23/04/26 15:39:57 WARN DAGScheduler: Creating new stage failed due to exception - job: 3
org.apache.spark.scheduler.BarrierJobUnsupportedRDDChainException: [SPARK-24820][SPARK-24821]: Barrier execution mode does not allow the following pattern of RDD chain within a barrier stage:
1. Ancestor RDDs that have different number of partitions from the resulting RDD (e.g. union()/coalesce()/first()/take()/PartitionPruningRDD). A workaround for first()/take() can be barrierRdd.collect().head (scala) or barrierRdd.collect()[0] (python).
2. An RDD that depends on multiple barrier RDDs (e.g. barrierRdd1.zip(barrierRdd2)).
        at org.apache.spark.errors.SparkCoreErrors$.barrierStageWithRDDChainPatternError(SparkCoreErrors.scala:225)
        at org.apache.spark.scheduler.DAGScheduler.checkBarrierStageWithRDDChainPattern(DAGScheduler.scala:486)
        at org.apache.spark.scheduler.DAGScheduler.createResultStage(DAGScheduler.scala:635)
        at org.apache.spark.scheduler.DAGScheduler.handleJobSubmitted(DAGScheduler.scala:1254)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2961)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2953)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2942)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
```

### Why are the changes needed?
To make barrier python UDF able to be supported in other dataframe operator in the future, in the mean time, do not expose it to the end users

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing UTs

### CODE REVIEW

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 1]
```diff
@@ -826,9 +826,6 @@ message MapPartitions {
 
   // (Required) Input user-defined function.
   CommonInlineUserDefinedFunction func = 2;
-
-  // (Optional) Whether to use barrier mode execution or not.
-  optional bool is_barrier = 3;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:831]: This is a breaking change right?

Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:831]: yes, but this field was added in master only. So it is not a breaking change against 3.4

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -333,6 +333,9 @@ message PythonUDF {
   bytes command = 3;
   // (Required) Python version being used in the client.
   string python_ver = 4;
+  // (Optional) Whether this PythonUDF should be executed in barrier mode.
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:336]: Since this barrier udf is only used to integrate other ML frameworks (pytorch for now, will support xgboost), so I guess it can be python only.

**Comment by zhengruifeng** [comment]
cc @WeichenXu123 @HyukjinKwon

File: python/pyspark/sql/udf.py [hunk 3]
```diff
@@ -465,6 +476,7 @@ def wrapper(*args: "ColumnOrName") -> Column:
         wrapper.returnType = self.returnType  # type: ignore[attr-defined]
         wrapper.evalType = self.evalType  # type: ignore[attr-defined]
         wrapper.deterministic = self.deterministic  # type: ignore[attr-defined]
+        wrapper.is_barrier = self.is_barrier  # type: ignore[attr-defined]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py:479]: I would keep this as an internal property by naming `_is_barrier`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala [hunk 4]
```diff
@@ -57,9 +57,18 @@ case class PythonUDF(
...
     evalType: Int,
     udfDeterministic: Boolean,
-    resultId: ExprId = NamedExpression.newExprId)
+    resultId: ExprId = NamedExpression.newExprId,
+    isBarrier: Boolean = false)
   extends Expression with Unevaluable with NonSQLExpression with UserDefinedExpression {
 
+  if (isBarrier &&
+    evalType != PythonEvalType.SQL_MAP_PANDAS_ITER_UDF &&
+    evalType != PythonEvalType.SQL_MAP_ARROW_ITER_UDF) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:66]:
```suggestion
      evalType != PythonEvalType.SQL_MAP_PANDAS_ITER_UDF &&
      evalType != PythonEvalType.SQL_MAP_ARROW_ITER_UDF) {
```

**Review by HyukjinKwon — DISMISSED** [review]
I support that we don't expose this to the end users. I am fine with this as a temporary solution.

File: python/pyspark/sql/udf.py (hunk 3 above)
Comment (zhengruifeng) [reply python/pyspark/sql/udf.py:479]: got it, let me rename it and add more comments

**Comment by WeichenXu123** [comment]
> This PR will not add a user-facing API or Parameter or Annotation, instead only a private function attribute will be added

Q: Where is related doc and example usage code for this attribute ? 

Q2: Why it is not a user-facing attribute ? We have 3rd-party library xgboost-spark that uses "barrier mode" and we need to make xgboost-spark supports pyspark-connect mode.

**Comment by zhengruifeng** [comment]
> Q: Where is related doc and example usage code for this attribute ?

I will add it.


> Q2: Why it is not a user-facing attribute ? We have 3rd-party library xgboost-spark that uses "barrier mode" and we need to make xgboost-spark supports pyspark-connect mode.

we can add doc and test for it, to avoid break it in the future.

But it is kind of `DeveloperApi`, our usage pattern is pretty limited. Moreover, it's likely to lead to weird behavior/failure to end users, due to the underlying `RDDBarrier` limitation.

**Comment by WeichenXu123** [comment]
> But it is kind of DeveloperApi, our usage pattern is pretty limited. Moreover, it's likely to lead to weird behavior/failure to end users, due to the underlying RDDBarrier limitation.

Our xgboost users are already aware of the limitations, it is not an issue.
note that currently xgboost library( python package) already uses RDD barrier API,
in future we need to adapt xgboost with spark connect mode, 
this means the SQL side barrier flag should also be an user-facing interface.

File: python/pyspark/sql/udf.py [hunk 5]
```diff
@@ -249,6 +259,38 @@ def __init__(
...
+        # in methods 'mapInPandas' and 'mapInArrow'.
+        # For example:
+        #
+        # df = spark.createDataFrame([(1, 21), (2, 30)], ("id", "age"))
+        #
+        # def filter_func(iterator):
+        #     for pdf in iterator:
+        #         yield pdf[pdf.id == 1]
+        #
+        # filter_func._is_barrier = True # Mark this UDF is barrier
```
Comment (zhengruifeng) [inline python/pyspark/sql/udf.py:275]: @WeichenXu123 I add an example here. also add UT to make sure it works.

> Our xgboost users are already aware of the limitations, it is not an issue.
> note that currently xgboost library( python package) already uses RDD barrier API,
> in future we need to adapt xgboost with spark connect mode,
> this means the SQL side barrier flag should also be an user-facing interface.

I'm aware of the integration with XGBoost. It looks like only developer will use it.  Is current implementation of a function attribute enough to support it?

The UDF is used much more widely than the RDDBarrier APIs, and my concern is that the end users is likely to abuse it.

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 2 above)
Comment (allisonwang-db) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:336]: Can you briefly describe in the comment what barrier mode is used for?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala [hunk 6]
```diff
@@ -73,9 +73,18 @@ case class PythonUDF(
     children: Seq[Expression],
     evalType: Int,
     udfDeterministic: Boolean,
-    resultId: ExprId = NamedExpression.newExprId)
+    resultId: ExprId = NamedExpression.newExprId,
+    isBarrier: Boolean = false)
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:77]: This seems to be very specific to the ML use case and not generic for Python UDFs. Do we expect more configs like this to be introduced in the future? Have we considered using a metadata or options map here?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto (hunk 2 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:336]: good point, will update. thx

File: python/pyspark/sql/udf.py (hunk 5 above)
Comment (WeichenXu123) [reply python/pyspark/sql/udf.py:275]: Define a `@barrier` decorator looks better, and we can document `barrier` doc saying this is a developer API and we should keep it stable

Comment (zhengruifeng) [reply python/pyspark/sql/udf.py:275]: will have a try

**Comment by zhengruifeng** [comment]
let me move the functions to sql.pandas.utils, will be less visible then

File: python/pyspark/sql/pandas/utils.py [hunk 7]
```diff
@@ -84,3 +87,57 @@ def pyarrow_version_less_than_minimum(minimum_pyarrow_version: str) -> bool:
         return False
 
     return LooseVersion(pyarrow.__version__) < LooseVersion(minimum_pyarrow_version)
+
+
+def barrier(f: Callable) -> Callable:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/utils.py:92]: If we're going to have this standalone API, this should work together with other similar API like `groupby().applyInPandas`.

**Comment by xinrong-meng** [comment]
It seems to me that functions under `python/pyspark/sql/pandas/utils.py` are used internally - in the existing Spark source code only.

A developer API "should" still be called externally, by developers though,  e.g. `semanticHash`.

So defining a developer API in `utils.py` seems a little unclear to me.

**Comment by WeichenXu123** [comment]
> It seems to me that functions under `python/pyspark/sql/pandas/utils.py` are used internally - in the existing Spark source code only.
> 
> A developer API "should" still be called externally, by developers though, e.g. `semanticHash`.
> 
> So defining a developer API in `utils.py` seems a little unclear to me.

Can we move the `barrier` API definition to proper place ?

**Comment by zhengruifeng** [comment]
> If we're going to have this standalone API, this should work together with other similar API like groupby().applyInPandas.

I think we won't support other Pandas API `groupby().applyInPandas`:

- it is non-trivial to support due to limitation of `RDDBarrier`;
- the ml side doesn't need them for now;


> It seems to me that functions under python/pyspark/sql/pandas/utils.py are used internally - in the existing Spark source code only.

> A developer API "should" still be called externally, by developers though, e.g. semanticHash.

That is a good point, what about keeping `barrier` in `pandas/utils.py` and only used it internally like other helper functions? 

@HyukjinKwon @WeichenXu123 @xinrong-meng

**Comment by WeichenXu123** [comment]
> It seems to me that functions under python/pyspark/sql/pandas/utils.py are used internally - in the existing Spark source code only.

This might be an issue, because the 3rd-party project "xgboost-spark" is going to use it. Currently "xgboost-spark" project already uses barrier RDD API, and in future we want to make it support spark connect.

**Comment by zhengruifeng** [comment]
I think we (@HyukjinKwon @WeichenXu123 and I) have reach to agreement that this PR (which introduces an `@barrier` annotation) is no longer, and we will keep current `barrier` parameter in `MapInPandas/Arrow`.

The reason is that barrier UDF is still too specific to `MapInPandas/Arrow`, due to the limitation of underlying `RDDBarrier`:

- can not use it within common dataframe operators like `select`, `withColumn`
- can not use it within other similar pandas function like `applyInPandas`;

I will close it in two days if no more comments.

also cc @mengxr

**Comment by xinrong-meng** [comment]
+1 for the proposal, thanks @zhengruifeng !

> I think we (@HyukjinKwon @WeichenXu123 and I) have reach to agreement that this PR (which introduces an `@barrier` annotation) is no longer, and we will keep current `barrier` parameter in `MapInPandas/Arrow`.
